From e4f7146c8e2794d73f1a8c47241cab54e7ef0e72 Mon Sep 17 00:00:00 2001
From: AURA Conversion Engine <aura@example.invalid>
Date: Tue, 16 Jun 2026 17:56:39 +0530
Subject: [PATCH 2/7] ASoC: codecs: lpass-va-macro: remove lpass-cdc macro
 framework dependency

Replace the downstream vendor lpass-cdc macro registration, event dispatch, and device-pointer indirection framework with the upstream direct component/platform registration model.

Runtime PM and clock helpers are stubbed for completion in subsequent patches that wire up the upstream clock framework and lpass-macro-common power domain helpers.

Signed-off-by: AURA Conversion Engine <aura@example.invalid>
---
 sound/soc/codecs/lpass-va-macro.c | 219 ++++++++----------------------
 1 file changed, 57 insertions(+), 162 deletions(-)

diff --git a/sound/soc/codecs/lpass-va-macro.c b/sound/soc/codecs/lpass-va-macro.c
index 0c5f994..5ad4eb2 100644
--- a/sound/soc/codecs/lpass-va-macro.c
+++ b/sound/soc/codecs/lpass-va-macro.c
@@ -15,7 +15,6 @@
 #include <sound/tlv.h>
 #include <linux/pm_runtime.h>
 #include <dsp/digital-cdc-rsc-mgr.h>
-#include "lpass-cdc.h"
 #include "lpass-cdc-clk-rsc.h"
 
 
@@ -131,6 +130,7 @@
 #define LPASS_CDC_VA_MACRO_SWR_STRING_LEN 80
 #define LPASS_CDC_VA_MACRO_CHILD_DEVICES_MAX 3
 #define LPASS_CDC_VA_MACRO_DEC_UNMUTE_DELAY_MS     10
+#define MIC_PAIR_MAX 4
 
 static const DECLARE_TLV_DB_SCALE(digital_gain, 0, 1, 0);
 static int va_tx_unmute_delay = LPASS_CDC_VA_TX_DMIC_UNMUTE_DELAY_MS;
@@ -283,13 +283,8 @@ static bool lpass_cdc_va_macro_get_data(struct snd_soc_component *component,
 			      struct lpass_cdc_va_macro_priv **va_priv,
 			      const char *func_name)
 {
-	*va_dev = lpass_cdc_get_device_ptr(component->dev, VA_MACRO);
-	if (!(*va_dev)) {
-		dev_err_ratelimited(component->dev,
-			"%s: null device for macro!\n", func_name);
-		return false;
-	}
-	*va_priv = dev_get_drvdata((*va_dev));
+	*va_dev = component->dev;
+	*va_priv = snd_soc_component_get_drvdata(component);
 	if (!(*va_priv) || !(*va_priv)->component) {
 		dev_err_ratelimited(component->dev,
 			"%s: priv is null for macro!\n", func_name);
@@ -316,6 +311,13 @@ static int lpass_cdc_va_macro_clk_div_get(struct snd_soc_component *component, u
 	return (int)va_priv->dmic_clk_div[mic_pair];
 }
 
+static int lpass_va_macro_tx_mclk_enable(struct snd_soc_component *component,
+					   bool enable)
+{
+	/* TODO: Patch-04 provides full implementation */
+	return 0;
+}
+
 static int lpass_cdc_va_macro_mclk_enable(
 			struct lpass_cdc_va_macro_priv *va_priv,
 			bool mclk_enable, bool dapm)
@@ -388,89 +390,6 @@ exit:
 	return ret;
 }
 
-static int lpass_cdc_va_macro_event_handler(struct snd_soc_component *component,
-				  u16 event, u32 data)
-{
-	struct device *va_dev = NULL;
-	struct lpass_cdc_va_macro_priv *va_priv = NULL;
-	int retry_cnt = MAX_RETRY_ATTEMPTS;
-	int ret = 0;
-
-	if (!lpass_cdc_va_macro_get_data(component, &va_dev,
-					 &va_priv, __func__))
-		return -EINVAL;
-
-	switch (event) {
-	case LPASS_CDC_MACRO_EVT_WAIT_VA_CLK_RESET:
-		while ((va_priv->va_mclk_users != 0) && (retry_cnt != 0)) {
-			dev_dbg_ratelimited(va_dev, "%s:retry_cnt: %d\n",
-				__func__, retry_cnt);
-			/*
-			 * Userspace takes 10 seconds to close
-			 * the session when pcm_start fails due to concurrency
-			 * with PDR/SSR. Loop and check every 20ms till 10
-			 * seconds for va_mclk user count to get reset to 0
-			 * which ensures userspace teardown is done and SSR
-			 * powerup seq can proceed.
-			 */
-			msleep(20);
-			retry_cnt--;
-		}
-		if (retry_cnt == 0)
-			dev_err_ratelimited(va_dev,
-				"%s: va_mclk_users non-zero, SSR fail!!\n",
-				__func__);
-		break;
-	case LPASS_CDC_MACRO_EVT_PRE_SSR_UP:
-		va_priv->pre_dev_up = true;
-		/* enable&disable VA_CORE_CLK to reset GFMUX reg */
-		ret = lpass_cdc_va_macro_core_vote(va_priv, true);
-		if (ret < 0) {
-			dev_err_ratelimited(va_priv->dev,
-				"%s: va request core vote failed\n",
-				__func__);
-			break;
-		}
-		ret = lpass_cdc_clk_rsc_request_clock(va_priv->dev,
-						va_priv->default_clk_id,
-						va_priv->clk_id, true);
-		if (ret < 0)
-			dev_err_ratelimited(va_priv->dev,
-				"%s, failed to enable clk, ret:%d\n",
-				__func__, ret);
-		else
-			lpass_cdc_clk_rsc_request_clock(va_priv->dev,
-						va_priv->default_clk_id,
-						va_priv->clk_id, false);
-		lpass_cdc_va_macro_core_vote(va_priv, false);
-		break;
-	case LPASS_CDC_MACRO_EVT_SSR_UP:
-		/* reset swr after ssr/pdr */
-		va_priv->reset_swr = true;
-		va_priv->dev_up = true;
-		break;
-	case LPASS_CDC_MACRO_EVT_CLK_RESET:
-		lpass_cdc_rsc_clk_reset(va_dev, VA_CORE_CLK);
-		break;
-	case LPASS_CDC_MACRO_EVT_SSR_DOWN:
-		va_priv->pre_dev_up = false;
-		va_priv->dev_up = false;
-		if ((!pm_runtime_enabled(va_dev) ||
-		     !pm_runtime_suspended(va_dev))) {
-			ret = lpass_cdc_runtime_suspend(va_dev);
-			if (!ret) {
-				pm_runtime_disable(va_dev);
-				pm_runtime_set_suspended(va_dev);
-				pm_runtime_enable(va_dev);
-			}
-		}
-		break;
-	default:
-		break;
-	}
-	return 0;
-}
-
 static int lpass_cdc_va_macro_mclk_event(struct snd_soc_dapm_widget *w,
 			       struct snd_kcontrol *kcontrol, int event)
 {
@@ -500,7 +419,7 @@ static int lpass_cdc_va_macro_mclk_event(struct snd_soc_dapm_widget *w,
 			if (va_priv->lpi_enable)
 				ret = lpass_cdc_va_macro_mclk_enable(va_priv, 1, true);
 			else
-				ret = lpass_cdc_tx_mclk_enable(component, 1);
+				ret = lpass_va_macro_tx_mclk_enable(component, true);
 		}
 		break;
 	case SND_SOC_DAPM_POST_PMD:
@@ -510,7 +429,7 @@ static int lpass_cdc_va_macro_mclk_event(struct snd_soc_dapm_widget *w,
 			if (va_priv->lpi_enable)
 				lpass_cdc_va_macro_mclk_enable(va_priv, 0, true);
 			else
-				lpass_cdc_tx_mclk_enable(component, 0);
+				lpass_va_macro_tx_mclk_enable(component, false);
 		}
 
 		if (va_priv->dapm_tx_clk_status > 0) {
@@ -1049,6 +968,13 @@ static int lpass_cdc_va_macro_tx_mixer_put(struct snd_kcontrol *kcontrol,
 	return 0;
 }
 
+static int lpass_va_macro_dmic_clk_enable(struct snd_soc_component *component,
+					 u32 dmic, bool enable)
+{
+	/* TODO: Patch-03 provides full implementation */
+	return 0;
+}
+
 static int lpass_cdc_va_macro_enable_dmic(struct snd_soc_dapm_widget *w,
 		struct snd_kcontrol *kcontrol, int event, u16 adc_mux0_cfg)
 {
@@ -1063,10 +989,10 @@ static int lpass_cdc_va_macro_enable_dmic(struct snd_soc_dapm_widget *w,
 
 	switch (event) {
 	case SND_SOC_DAPM_PRE_PMU:
-		lpass_cdc_dmic_clk_enable(component, (u32)dmic, (u32)DMIC_VA, true);
+		lpass_va_macro_dmic_clk_enable(component, (u32)dmic, true);
 		break;
 	case SND_SOC_DAPM_POST_PMD:
-		lpass_cdc_dmic_clk_enable(component, (u32)dmic, (u32)DMIC_VA, false);
+		lpass_va_macro_dmic_clk_enable(component, (u32)dmic, false);
 		break;
 	}
 
@@ -2049,17 +1975,11 @@ static int lpass_cdc_va_macro_init(struct snd_soc_component *component)
 {
 	struct snd_soc_dapm_context *dapm =
 				snd_soc_component_to_dapm(component);
-	int ret, i;
-	struct device *va_dev = NULL;
-	struct lpass_cdc_va_macro_priv *va_priv = NULL;
+	int i;
+	struct device *va_dev = component->dev;
+	struct lpass_cdc_va_macro_priv *va_priv;
 
-	va_dev = lpass_cdc_get_device_ptr(component->dev, VA_MACRO);
-	if (!va_dev) {
-		dev_err(component->dev,
-			"%s: null device for macro!\n", __func__);
-		return -EINVAL;
-	}
-	va_priv = dev_get_drvdata(va_dev);
+	va_priv = snd_soc_component_get_drvdata(component);
 	if (!va_priv) {
 		dev_err(component->dev,
 			"%s: priv is null for macro!\n", __func__);
@@ -2070,40 +1990,6 @@ static int lpass_cdc_va_macro_init(struct snd_soc_component *component)
 	va_priv->swr_dmic_enable = false;
 	//va_priv->register_event_listener = false;
 
-	va_priv->version = lpass_cdc_get_version(va_dev);
-
-	ret = snd_soc_dapm_new_controls(dapm,
-			lpass_cdc_va_macro_dapm_widgets,
-			ARRAY_SIZE(lpass_cdc_va_macro_dapm_widgets));
-	if (ret < 0) {
-		dev_err(va_dev, "%s: Failed to add controls\n",
-			__func__);
-		return ret;
-	}
-
-	ret = snd_soc_dapm_add_routes(dapm, va_audio_map,
-				ARRAY_SIZE(va_audio_map));
-	if (ret < 0) {
-		dev_err(va_dev, "%s: Failed to add routes\n",
-			__func__);
-		return ret;
-	}
-
-	ret = snd_soc_dapm_new_widgets(snd_soc_dapm_to_card(dapm));
-	if (ret < 0) {
-		dev_err(va_dev, "%s: Failed to add widgets\n", __func__);
-		return ret;
-	}
-
-	ret = snd_soc_add_component_controls(component,
-			lpass_cdc_va_macro_snd_controls,
-			ARRAY_SIZE(lpass_cdc_va_macro_snd_controls));
-	if (ret < 0) {
-		dev_err(va_dev, "%s: Failed to add snd_ctls\n",
-			__func__);
-		return ret;
-	}
-
 	snd_soc_dapm_ignore_suspend(dapm, "VA_AIF1 Capture");
 	snd_soc_dapm_ignore_suspend(dapm, "VA_AIF2 Capture");
 	snd_soc_dapm_ignore_suspend(dapm, "VA_AIF3 Capture");
@@ -2246,22 +2132,19 @@ err:
 	return;
 }
 
-static void lpass_cdc_va_macro_init_ops(struct macro_ops *ops,
-			      char __iomem *va_io_base)
-{
-	memset(ops, 0, sizeof(struct macro_ops));
-	ops->dai_ptr = lpass_cdc_va_macro_dai;
-	ops->num_dais = ARRAY_SIZE(lpass_cdc_va_macro_dai);
-	ops->init = lpass_cdc_va_macro_init;
-	ops->exit = lpass_cdc_va_macro_deinit;
-	ops->io_base = va_io_base;
-	ops->event_handler = lpass_cdc_va_macro_event_handler;
-	ops->clk_div_get = lpass_cdc_va_macro_clk_div_get;
-}
+static const struct snd_soc_component_driver lpass_cdc_va_macro_component_drv = {
+	.name = "VA MACRO",
+	.probe = lpass_cdc_va_macro_init,
+	.controls = lpass_cdc_va_macro_snd_controls,
+	.num_controls = ARRAY_SIZE(lpass_cdc_va_macro_snd_controls),
+	.dapm_widgets = lpass_cdc_va_macro_dapm_widgets,
+	.num_dapm_widgets = ARRAY_SIZE(lpass_cdc_va_macro_dapm_widgets),
+	.dapm_routes = va_audio_map,
+	.num_dapm_routes = ARRAY_SIZE(va_audio_map),
+};
 
 static int lpass_cdc_va_macro_probe(struct platform_device *pdev)
 {
-	struct macro_ops ops;
 	struct lpass_cdc_va_macro_priv *va_priv;
 	u32 va_base_addr, prop_size, *temp;
 	char __iomem *va_io_base;
@@ -2438,13 +2321,14 @@ static int lpass_cdc_va_macro_probe(struct platform_device *pdev)
 	mutex_init(&va_priv->mclk_lock);
 	mutex_init(&va_priv->wlock);
 	dev_set_drvdata(&pdev->dev, va_priv);
-	lpass_cdc_va_macro_init_ops(&ops, va_io_base);
-	ops.clk_id_req = va_priv->default_clk_id;
-	ops.default_clk_id = va_priv->default_clk_id;
-	ret = lpass_cdc_register_macro(&pdev->dev, VA_MACRO, &ops);
+	ret = devm_snd_soc_register_component(&pdev->dev,
+			&lpass_cdc_va_macro_component_drv,
+			lpass_cdc_va_macro_dai,
+			ARRAY_SIZE(lpass_cdc_va_macro_dai));
 	if (ret < 0) {
-		dev_err(&pdev->dev, "%s: register macro failed\n", __func__);
-		goto reg_macro_fail;
+		dev_err(&pdev->dev, "%s: register component failed
+", __func__);
+		goto reg_component_fail;
 	}
 	pm_runtime_set_autosuspend_delay(&pdev->dev, VA_AUTO_SUSPEND_DELAY);
 	pm_runtime_use_autosuspend(&pdev->dev);
@@ -2455,7 +2339,7 @@ static int lpass_cdc_va_macro_probe(struct platform_device *pdev)
 		schedule_work(&va_priv->lpass_cdc_va_macro_add_child_devices_work);
 	return ret;
 
-reg_macro_fail:
+reg_component_fail:
 	mutex_destroy(&va_priv->mclk_lock);
 	mutex_destroy(&va_priv->wlock);
 	if (is_used_va_swr_gpio)
@@ -2483,7 +2367,6 @@ static void lpass_cdc_va_macro_remove(struct platform_device *pdev)
 		&va_priv->va_dec_unmute_work.dwork);
 	pm_runtime_disable(&pdev->dev);
 	pm_runtime_set_suspended(&pdev->dev);
-	lpass_cdc_unregister_macro(&pdev->dev, VA_MACRO);
 	mutex_destroy(&va_priv->mclk_lock);
 	mutex_destroy(&va_priv->wlock);
 	if (va_priv->is_used_va_swr_gpio)
@@ -2496,14 +2379,26 @@ static const struct of_device_id lpass_cdc_va_macro_dt_match[] = {
 	{}
 };
 
+static int va_macro_runtime_suspend(struct device *dev)
+{
+	/* TODO: Patch-04 provides full implementation */
+	return 0;
+}
+
+static int va_macro_runtime_resume(struct device *dev)
+{
+	/* TODO: Patch-04 provides full implementation */
+	return 0;
+}
+
 static const struct dev_pm_ops lpass_cdc_dev_pm_ops = {
 	SET_SYSTEM_SLEEP_PM_OPS(
 		pm_runtime_force_suspend,
 		pm_runtime_force_resume
 	)
 	SET_RUNTIME_PM_OPS(
-		lpass_cdc_runtime_suspend,
-		lpass_cdc_runtime_resume,
+		va_macro_runtime_suspend,
+		va_macro_runtime_resume,
 		NULL
 	)
 };
-- 
2.34.1

